src/panel/applets/status/PowerProfiles.vala

```diff
@@ -0,0 +1,68 @@
+/*
+ * This file is part of budgie-desktop
+ *
+ * Copyright Budgie Desktop Developers
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ */
+
+using GLib;
+
+[DBus (name = "net.hadess.PowerProfiles")]
+public interface PowerProfilesDBus : GLib.Object {
+
+	[DBus (name = "HoldProfile")]
+	public abstract uint hold_profile(string profile, string reason, string application_id) throws DBusError, IOError;
+
+	[DBus (name = "ReleaseProfile")]
+	public abstract void release_profile(uint cookie) throws DBusError, IOError;
+
+	[DBus (name = "ProfileReleased")]
+	public signal void profile_released(uint cookie);
+
+	[DBus (name = "ActiveProfile")]
+	public abstract string active_profile { owned get; set; }
+
+	[DBus (name = "PerformanceInhibited")]
+	public abstract string performance_inhibited { owned get; }
+
+	[DBus (name = "PerformanceDegraded")]
+	public abstract string performance_degraded { owned get; }
+
+	[DBus (name = "Profiles")]
+	public abstract GLib.HashTable<string, GLib.Variant>[] profiles { owned get; }
+
+	[DBus (name = "Actions")]
+	public abstract string[] actions { owned get; }
+
+	[DBus (name = "ActiveProfileHolds")]
+	public abstract GLib.HashTable<string, GLib.Variant>[] active_profile_holds { owned get; }
+}
+ 
+public class PowerProfiles : GLib.Object {
+	public signal void proxy_connected ();
+
+	public PowerProfilesDBus? proxy = null;
+
+	public PowerProfiles() {
+		get_proxy();
+	}
+
+	public void get_proxy() {
+		if (proxy == null) {
+			Bus.get_proxy.begin<PowerProfilesDBus>(BusType.SYSTEM, "net.hadess.PowerProfiles", "/net/hadess/PowerProfiles", 0, null, on_proxy_get);
+		}
+	}
+
+	void on_proxy_get(Object? o, AsyncResult? res) {
+		try {
+			proxy = Bus.get_proxy.end(res);
+			proxy_connected();
+		} catch (Error e) {
+			warning("unable to connect to net.hadess.PowerProfiles- power-profiles-daemon is probably not installed. %s", e.message);
+		}
+	}
+}
```

Comment:
I'm not sure this little class is necessary. I would suggest moving the `Bus.get_proxy` call to near the end of the constructor of PowerIndicator, and moving the `on_proxy_get` function to that class. In that function, you can then connect your signals and do the initial state setting.